### PR TITLE
Connecttrace fewer blocks

### PR DESCRIPTION
**Pull request by narula**
During a large reorg, the number of blocks kept in memory could grow to be very large.  This change makes it so that only the last 10 blocks are kept in memory, and if necessary, older blocks are re-read from disk.

Also introduce a parameter to FlushStateToDisk to ensure that we do not try to prune during the life of a ConnectTrace, because we might need to re-read the block from disk.

Addresses https://github.com/bitcoin/bitcoin/issues/9027

### CODE REVIEW

**Comment by jonasschnelli** [comment]
Concept ACK (not looked to closely at the code yet).
10 blocks seems to be little... maybe it should be configurable (most memory consuming processes are configurable right now).

**Comment by gmaxwell** [comment]
why would ten be too little?  If it made the code simpler I would have expected it to just do one, reading them isn't free but it's not that expensive-- shouldn't be the limiting factor in reorg performance.

**Comment by jonasschnelli** [comment]
> why would ten be too little?

I kinda followed the assumption that some users want to reorg as fast as possible (I's probably often a tiny difference) regardless of the depth of the reorganisation. I guess those users do not care about using 200MB+ for tracing the blocks in mem during a reorg. But it's a wild assumptions and maybe I'm wrong.

**Comment by TheBlueMatt** [comment]
Well I'm not sure how many users are really focused on how long a 20 block reorg will take...should be sufficiently rare that a user spending a lot of time thinking about it is probably less than useful.

**Comment by gmaxwell** [comment]
@jonasschnelli reorgs are slow but for reasons that have nothing to do with reading blocks from the disk.  it can easily take many seconds to reorg 20 blocks, reading them will take a few milliseconds.  Making reorgs faster is great but I think this is the wrong place to optimize for that.

File: src/validation.cpp [hunk 1]
```diff
@@ -186,7 +186,7 @@ enum FlushStateMode {
 };
 
 // See definition for documentation
-static bool FlushStateToDisk(const CChainParams& chainParams, CValidationState &state, FlushStateMode mode, int nManualPruneHeight=0);
+static bool FlushStateToDisk(const CChainParams& chainParams, CValidationState &state, FlushStateMode mode, int nManualPruneHeight, bool prune);
```
Comment (sipa) [inline src/validation.cpp]: It seems the `prune` flag is implied by `mode != FLUSH_STATE_IS_NEEDED`.

Comment (narula) [inline src/validation.cpp]: Yes, you're right.  But is it reasonable to rely on that for prune's semantics?

File: src/validation.cpp [hunk 2]
```diff
@@ -2029,6 +2032,12 @@ class ConnectTrace {
         assert(pblock);
         blocksConnected.back().pindex = pindex;
         blocksConnected.back().pblock = std::move(pblock);
+        if (++num_connected >= MAX_CONNECT_TRACE_BLOCKS_IN_MEMORY) {
```
Comment (JeremyRubin) [inline src/validation.cpp]: slight preference for 

```c++
++num_connected;
if (num_connected >= MAX_CONNECT_TRACE_BLOCKS_IN_MEMORY)
```

File: src/validation.cpp [hunk 3]
```diff
@@ -2336,8 +2345,18 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
             fInitialDownload = IsInitialBlockDownload();
 
             for (const PerBlockConnectTrace& trace : connectTrace.GetBlocksConnected()) {
-                assert(trace.pblock && trace.pindex);
-                GetMainSignals().BlockConnected(trace.pblock, trace.pindex, *trace.conflictedTxs);
+                assert(trace.pindex);
+                std::shared_ptr<const CBlock> pthisBlock;
```
Comment (JeremyRubin) [inline src/validation.cpp]: probably a bit cleaner to just do:
```c++
std::shared_ptr<const CBlock> pthisBlock(trace.pblock);
if (!pThisBlock) {

}
```

File: src/validation.cpp [hunk 4]
```diff
@@ -2336,8 +2345,18 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
...
+                if (trace.pblock == NULL) {
+                    std::shared_ptr<CBlock> pblockNew = std::make_shared<CBlock>();
```
Comment (JeremyRubin) [inline src/validation.cpp]: probably no need to make pblockNew? Can just assign directly with make_shared to pthisBlock.

**Review by JeremyRubin — COMMENTED** [review]
Generally, looks ok to me, a couple style nits.

I think the one thing that I would want to consider is that counting the number of blocks is not an exact match for the amount of memory used. Maybe you can count blocks based on their memory weight. For example, if we were to ever add some really big data structure in memory when we load a block this code would break, but a limit in MB would just be more aggressive in evicting from memory.

I also think that it might be nice if you can maybe increase the batching or do an async read. Because now it seems that the disk seeks are now in the hot path, whereas it might be possible to prefetch them and never wait. Then again, this case (> 10 reorg) is rare enough it's probably premature to care that much about latency here.

**Pull request closed without merging**